### DISCUSSION TITLE
exposing external crate's (APIs) to Lua world (as a module)

### DISCUSSION THREAD

**Discussion by ildar** [mlua-rs/mlua#117] [Q&A] [answered]
Greetings!
I'm discovering how to expose a Rust external crate as a Lua module.
I've already stumbled on a few things so wanted to ask community for a good advise.
1. to store some structures in Lua tables (as UserData) they should be `Clone` and `Copy`. If not, you're probably in a big trouble.
2. if you cannot store something in a Lua table (see above) you need some static storage for it, right? Maybe something with `RefCell<Option<Data>>`? That's a huge antipattern in Rust and I'm a little lost on how to handle it in a right way.

P.S. There are definitely patterns and antipatterns for this. Anyone wants to share?
P.P.S. I've taken a quick look at other projects and couldn't find a good example for the topic. Anyone?

**Comment by khvzak** [comment]
> to store some structures in Lua tables (as UserData) they should be Clone and Copy. If not, you're probably in a big trouble.

Hm, there is no such requirement. Any data structure can be passed to Lua without implementing `Clone` and/or `Copy`.
See [an example](https://github.com/khvzak/mlua/blob/master/examples/userdata.rs) where `Rectangle` struct implements neither clone nor copy.

**Comment by ildar** [comment]
Partially.
But I fail to access it:
```
   |
33 |     let test: Result<&MyUserData,LuaError> = exports.get("userdata");
   |                                                      ^^^ the trait
`mlua::UserData` is not implemented for `&MyUserData`
```
The `get()` getter returns the copy of the stored object. I see no way to get the ref.
[full source link](https://paste.rs/b98.rs)

**Comment by khvzak** [comment, marked as answer]
This is because you need to use [`AnyUserData`](https://docs.rs/mlua/0.7.1/mlua/struct.AnyUserData.html) instead of `&MyUserData` to get your data from the table.
And then borrow it using [`borrow()`](https://docs.rs/mlua/0.7.1/mlua/struct.AnyUserData.html#method.borrow) or `borrow_mut` methods.
Eg:
```rust
let ud = exports.get::<_, AnyUserData>("userdata")?;
let mydata = ud.borrow::<MyUserData>()?;
```

**Comment by ildar** [comment]
Just what I need! Thanks.
BTW, do you know other "expose a Rust external crate as a Lua module" examples? I found https://crates.io/crates/distant-lua but it uses mlua 0.6 and is relatively simple.

**Comment by ildar** [comment]
hmm, `borrow()` implies `'static` lifetime to UserData contents which isn't
always true, I think. Need to elaborate that. Meanwhile JFYI, I'm stumbling
on the error:
```
error[E0621]: explicit lifetime required in parameter type
  --> src/lib.rs:37:25
   |
37 |         let mut el = el.borrow_mut::<EventLoop>()?;
   |                         ^^^^^^^^^^ lifetime `'static` required
```
full project: https://github.com/ildar/lua-module-calloop

**Reply by khvzak** [reply]
`'static` lifetime is Rust requirement for `Any` trait: https://doc.rust-lang.org/std/any/trait.Any.html

Alternatively you can use non-static userdata types inside [scope](https://docs.rs/mlua/latest/mlua/struct.Lua.html#method.scope).

**Reply by ildar** [reply]
This is again more complicated then I thought.
I've made a small drawing, I wonder if I'm getting it all right.
![mlua_userdata](https://user-images.githubusercontent.com/517752/147968477-244f7e55-8d59-478a-9072-5cadd58ae8d4.png)
[source](https://drive.google.com/file/d/1KBruvZhzXbgsmMvmLsAek_UPROpy5-64/view)

**Reply by khvzak** [reply]
Yes, the diagram above is almost correct.

`Clone` for `UserData` is requirement **only** for [`FromLua`](https://github.com/khvzak/mlua/blob/master/src/conversion.rs#L148) implementation.

non-`'static` types are very tricky because they don't implement `Any` trait (dynamic typing). In Lua all types information is literally erased and metamethods used to match to a particular Rust type. For `'static` types all userdata objects share a single metatable, for non-`'static` inside scope all instances have uniq metatables. 

If it's possible I would recommend either go to `unsafe {}` territory to make things much easier or lift lifetimes to `'static`.

**Comment by ildar** [comment]
Surprisingly, I managed to skip using `scope()`. After some meditation and
a few experiments I declared exported function's argument as 'static:
https://github.com/ildar/lua-module-calloop/commit/a57ff0b944d9e9a779e6c8912ca5cf5a9b5300a6#diff-b1a35a68f14e696205874893c07fd24fdb88882b47c23cc0e0c80a30c7d53759R34
Builds and works for me.
